File: src/main/java/com/josephbaca/world/World.java
package com.josephbaca.world;

import com.josephbaca.rpggame.ContextManager;

/**
 * Contains all the geography about the world the player is in. Specifically,
 * it contains the coordinates of every {@link Tile} and the current location
 * of the player.
 */
public class World {

  private static final org.slf4j.Logger LOG = org.slf4j.LoggerFactory.getLogger(World.class);

  private final String name;

  /**
   * Player coordinates.
   */
  private Coordinate playerCoords;

  private final ContextManager context;
  /**
   * Internal representation of the world.
   */
  private CoordinateGrid<Room> grid;

  /**
   * Creates a world.
   *
   * @param name of the world
   * @param maxx coordinate of the world
   * @param maxy coordinate of the world
   */
  public World(String name, int maxx, int maxy, ContextManager context) {
    this.name = name;

    this.grid = new CoordinateGrid<>(maxx, maxy);
    this.playerCoords = Coordinate.of(0, 0);

    generateWorld();

    this.context = context;
    context.addContextLayer(getCurrentRoom()); // Add current room to context

    LOG.info("Created world logger");
  }

  private void setPlayerCoords(Coordinate c) {
    if (c.x >= 0 && c.x < grid.sizeX() && c.y > 0 && c.y < grid.sizeY()) {
      playerCoords = c;
      context.replaceContextLayer(getCurrentRoom());
    } else {
      LOG.info("Player tried to walk off the world.");
    }
  }

  /**
   * Move the player up one unit.
   */
  public void movePlayerUp() {
    setPlayerCoords(Coordinate.of(playerCoords.x, playerCoords.y + 1));
  }

  /**
   * Move the player down one unit.
   */
  public void movePlayerDown() {
    setPlayerCoords(Coordinate.of(playerCoords.x, playerCoords.y - 1));
  }

  /**
   * Move the player right one unit.
   */
  public void movePlayerRight() {
    setPlayerCoords(Coordinate.of(playerCoords.x + 1, playerCoords.y));
  }

  /**
   * Move the player left one unit.
   */
  public void movePlayerLeft() {
    setPlayerCoords(Coordinate.of(playerCoords.x - 1, playerCoords.y));
  }

  private Room getCurrentRoom() {
    return grid.getCoordinate(playerCoords);
  }

  /**
   * World generator function.
   */
  private void generateWorld() {
    for (int x = 0; x < grid.sizeX(); x++) {
      for (int y = 0; y < grid.sizeY(); y++) {
        Room r = new Room(5, 5, this);
        grid.setCoordinate(Coordinate.of(x, y), r);
      }
    }
  }

  @Override
  public String toString() {
    return grid.toString();
  }

  /**
   * Returns a print friendly version of the world.
   */
  public String toDisplayString() {
    return grid.toDisplayString();
  }
}

File: src/main/java/com/josephbaca/world/World.kt
package com.josephbaca.world

import com.josephbaca.rpggame.GameStateManager
import com.josephbaca.context.Room
import com.josephbaca.context.RoomNouns

/**
 * Contains all the geography about the world the player is in. Specifically,
 * it contains the coordinates of every [Room] and the current location
 * of the player.
 */
class World(maxx: Int, maxy: Int, private val gameStateManager: GameStateManager) {

    private val worldGrid: CoordinateGrid<Room> = CoordinateGrid(maxx, maxy)

    private var playerCoordinates: Coordinate = Coordinate(0, 0)
        set(value) {
            if (value.x in 0 until worldGrid.sizeX && value.y in 0 until worldGrid.sizeY) {
                LOG.info(
                    "Player moving from (%s, %s) to (%s, %s)"
                        .format(playerCoordinates.x, playerCoordinates.y, value.x, value.y)
                )
                field = value
                gameStateManager.replaceContextLayer(currentRoom)
            } else {
                LOG.info(
                    "Player tried to walk off the world. Still at (%s, %s)."
                        .format(playerCoordinates.x, playerCoordinates.y)
                )
            }
        }

    var playerOrientation: Directions = Directions.NORTH

    private val currentRoom: Room
        get() = worldGrid.getCoordinate(playerCoordinates)

    init {
        generateWorld()
        gameStateManager.addContextLayer(currentRoom) // Add current room to gameStateManager
        LOG.debug("Added current room (%s) to context.".format(currentRoom))
        LOG.debug("Context is: %s".format(gameStateManager.contextStack))
    }

    /**
     * World generator function.
     */
    private fun generateWorld() {
        for (x in 0 until worldGrid.sizeX) {
            for (y in 0 until worldGrid.sizeY) {
                val r = Room(gameStateManager)
                worldGrid.setCoordinate(Coordinate(x, y), r)
            }
        }
    }

    fun movePlayer(relativeDirection: RoomNouns): String? {

        val oldPlayerOrientation = playerOrientation
        val oldPlayerCoords = playerCoordinates

        playerOrientation = getNewPlayerOrientation(relativeDirection) ?: return null
        playerCoordinates = getNewPlayerCoords()

        return if (oldPlayerCoords == playerCoordinates) {
            playerOrientation = oldPlayerOrientation
            "Oh heck! Sorry there's a wall there..."
        } else {
            "Okay cool you went that way!"
        }
    }

    private fun getNewPlayerOrientation(relativeDirection: RoomNouns): World.Directions? {

        return when (relativeDirection) {
            RoomNouns.FORWARD -> playerOrientation
            RoomNouns.BACK -> playerOrientation.back()
            RoomNouns.RIGHT -> playerOrientation.right()
            RoomNouns.LEFT -> playerOrientation.left()
            else -> return null // Didn't pass a direction
        }
    }

    private fun getNewPlayerCoords(): Coordinate {
        return when (playerOrientation) {
            Directions.NORTH -> Coordinate(playerCoordinates.x, playerCoordinates.y + 1)
            Directions.SOUTH -> Coordinate(playerCoordinates.x, playerCoordinates.y - 1)
            Directions.EAST -> Coordinate(playerCoordinates.x + 1, playerCoordinates.y)
            Directions.WEST -> Coordinate(playerCoordinates.x - 1, playerCoordinates.y)
        }
    }

    override fun toString(): String {
        return "World (%sx%s)".format(worldGrid.sizeX, worldGrid.sizeY)
    }

    enum class Directions {
        EAST,
        WEST,
        NORTH,
        SOUTH;

        fun back(): Directions {
            return when (this) {
                EAST -> WEST
                WEST -> EAST
                NORTH -> SOUTH
                SOUTH -> NORTH
            }
        }

        fun right(): Directions {
            return when (this) {
                EAST -> SOUTH
                WEST -> NORTH
                NORTH -> EAST
                SOUTH -> WEST
            }
        }

        fun left(): Directions {
            return when (this) {
                EAST -> NORTH
                WEST -> SOUTH
                NORTH -> WEST
                SOUTH -> EAST
            }
        }
    }

    companion object {

        private val LOG = org.slf4j.LoggerFactory.getLogger(World::class.java)
    }
}
